field/composite.go

```diff
@@ -0,0 +1,325 @@
+	"fmt"
+	"reflect"
+	"sort"
+	"strconv"
+
+	"github.com/moov-io/iso8583/padding"
+)
+
+var _ Field = (*Composite)(nil)
+
+// Composite is a wrapper object designed to hold ISO8583 subfields and
+// subelements.  Because Composite handles both of these usecases generically,
+// we refer to them collectively as 'subfields' throughout the receiver's
+// documentation and error messages.
+//
+// Composite handles aggregate fields of the following format:
+// - Length (if variable)
+// - []Subfield
+//
+// Where the subfield structure is assumed to be as follows:
+// - Subfield ID (if Composite.Spec().IDLength > 0)
+// - Subfield Length (if variable)
+// - Subfield data
+//
+// Composite behaves in two modes depending on whether subfield IDs need to be
+// explicitly handled or not. This is configured by setting Spec.IDLength.
+//
+// When populated, Composite handles the packing and unpacking subfield IDs on
+// their behalf. However, responsibility for packing and unpacking both the
+// length and data of subfields is delegated to the subfields themselves.
+// Therefore, their specs should be configured to handle such behavior.
+//
+// If Spec.IDLength > 0, Composite leverages Spec.Enc to unpack subfields
+// regardless of order based on their IDs. Similarly, it is also able to handle
+// non-present subfields by relying on the existence of their IDs.
+//
+// If Spec.IDLength == 0, Composite only unpacks subfields ordered by ID. The absence
+// of IDs in the data means that the receiver is not able to handle non-present
+// subfields either.
+//
+// For the sake of determinism, packing of subfields is executed in order of ID
+// regardless of the value of Spec.IDLength.
+//
+// Padding is not supported by Composite. Responsibility for this is delegated
+// recursively to the subfields themselves.
+type Composite struct {
+	spec *Spec
+
+	orderedSpecFieldIDs []int
+	idToFieldMap        map[int]Field
+
+	fieldsMap map[int]struct{}
+	data      *reflect.Value
+}
+
+// NewComposite creates a new instance of the *Composite struct,
+// validates and sets its Spec before returning it.
+// Refer to SetSpec() for more information on Spec validation.
+func NewComposite(spec *Spec) *Composite {
+	f := &Composite{
+		fieldsMap: map[int]struct{}{},
+	}
+	f.SetSpec(spec)
+	return f
+}
+
+// Spec returns the receiver's spec.
+func (f *Composite) Spec() *Spec {
+	return f.spec
+}
+
+// SetSpec validates the spec and creates new instances of Fields defined
+// in the specification.
+// NOTE: Composite does not support padding. Therefore, users should
+// only pass None or nil values for ths type. Passing any other value will
+// result in a panic.
+func (f *Composite) SetSpec(spec *Spec) {
+	if err := validateCompositeSpec(spec); err != nil {
+		panic(err)
+	}
+	f.spec = spec
+	f.idToFieldMap = spec.CreateMessageFields()
+	f.orderedSpecFieldIDs = orderedKeys(f.idToFieldMap)
+}
+
+// SetData traverses through fields provided in the data parameter matches them
```

Comment:
I see that you moved code that sets data into `pack` method. It's more compact now, but in my opinion it has following drawbacks:
* if there is an error for setting data (e.g. wrong type) you will know about it during `pack`, not `SetData`
* when you call `SetData` you may want to access message fields in the struct, but values are not there until you call `pack`

Is there a reason to move the logic described in the comment into `pack` method?

Reply:
Yes, I moved the code into both `Pack` and `Unpack` from `SetData`. I've made the same changes in `message.go`.

You're right about the type checking drawback. Unfortunately, the issue we had previously with `SetData` was that it used reflection to perform checks and created new instances of all nil fields but we needed more granularity on when to initialise these new fields.

When we are packing fields, we actually never wish to initialise nil fields because it is up to the user to decide which fields get converted to bytes.

When unpacking fields, we wish to initialise the relevant fields whose data has been deserialised from bytes (leaving non-present fields as nil).

We've added a test to assert against this in `composite_test.go`: `"Unpack correctly deserialises partial subfields to the data struct"` line 449 where we test against a struct which should be partially populated.

This test previously failed because `F2` was populated with its default value (empty string) instead of being nil and what led me to refactoring its implementation.

Because `SetData` is shared between both the packing and unpacking usecases, we aren't able to make fine grained decisions on what type checks to make and which fields to initialise and as such we are forced to move that logic into `Pack()` and `Unpack()`

> when you call SetData you may want to access message fields in the struct, but values are not there until you call pack

I haven't quite understood this dawback though. Users should be able to access the fields of their data struct directly because they created it and may hold a reference to its pointer if they desire.

In addition, calling `String()` or `Bytes()` calls `pack()` under the hood so they should be able to access these fields via the untyped API. Is there an entry point that I'm missing out on?